File: Algo/Leetcode/671SecondMinInBinaryTree.cpp
class Solution {
public:
    int res = -1;
    int findSecondMinimumValue(TreeNode* root) {
        if(!root) return -1;
        if(!root->left || !root->right) return -1;
        dfs(root, root->val);
        return res;
    }

    void dfs(TreeNode* root, int m){
        if(!root) return;
        if(root->val > m && (root->val < res || res == -1))
            res = root->val;
        dfs(root->left, m);
        dfs(root->right, m);
    }
};

File: Algo/Leetcode/671SecondMinInBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def __init__(self):
        self.minV = 0
        self.res = sys.maxint
        self.find = False

    def findSecondMinimumValue(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root is None:
            return -1
        self.minV = root.val
        self.dfs(root)

        return self.res if self.find else -1

    def dfs(self, root):
        if root is None:
            return

        self.dfs(root.left)
        if root.val > self.minV and root.val < self.res:
            self.find = True
            self.res = root.val
        self.dfs(root.right)